SciTools/iris
# Replace pyke

## 🚀 Pull Request

### Description
Eventually, will address #3415

#### Technical details of approach
  * I started by writing a drop-in replacement approach to replace Pyke rules with so-called 'action' routines.
    * at present it purposely mimics the structure of existing Pyke rules, plus the three-stage translation process built into 
       `iris.fileformats.netcdf._load_cube` , i.e. broadly [assert facts](https://github.com/SciTools/iris/blob/master/lib/iris/fileformats/netcdf.py#L592-L604); [run rules](https://github.com/SciTools/iris/blob/master/lib/iris/fileformats/netcdf.py#L592-L604); [complete cube](https://github.com/SciTools/iris/blob/master/lib/iris/fileformats/netcdf.py#L609-L652)
    * it does that by providing a new [thin wrapper object](https://github.com/SciTools/iris/blob/b118666e62a194eb131513ce77aeef869beba87b/lib/iris/fileformats/_nc_load_rules/engine.py#L57-L64) which can [mimic the Pyke "engine"](https://github.com/SciTools/iris/blob/b118666e62a194eb131513ce77aeef869beba87b/lib/iris/fileformats/netcdf.py#L418-L430).
    * for now, I have retained the current approach to storing Iris info on the 'engine' ([example](https://github.com/SciTools/iris/blob/22ff3ebacf04b741994578489706a833a3f05fb6/lib/iris/fileformats/_nc_load_rules/actions.py#L181)), and also [mimicked the storage of intermediate 'facts' there too](https://github.com/SciTools/iris/blob/22ff3ebacf04b741994578489706a833a3f05fb6/lib/iris/fileformats/_nc_load_rules/engine.py#L91-L98).
    * some of that could well simplify in future, but I'm aiming to ***first ensure equivalence with the exisiting behaviour***
    * N.B. currently incomplete, have only so far done the 'major' rules (grid-mapping and dim-coords).
  * I am also creating some targetted testing for this ...
    * the choice of tests is heavily informed by the structure of the existing rules
    * the area tested, essentially what is currently "the Pyke rules bit", is not maybe _that_ obvious a separate distinct category ..
      * .. but it should be an appropriate starting point for testing the 'behavioural equivalence'
      * I intend eventually to test for each code-branch of the rules behaviour
    * ... so, again, in future we could simplify/rework this 

#### Current progress + status summary
Done
  * actions + tests for
    * grid-mappings + dim-coords
    * time-coords
    * hybrid vertical coords

Todo:
  * some areas are still unaddressed - lacking _both_ actions implementation + tests
  * known problems
    1. already observed that actions do **_not_**  match original behaviour for some of the grid-mapping/dimcoords areas
         * ( intend to complete the test coverage before addressing this )
    2. missing parts of grid-mapping/dimcoords tests : 
        * still not exercising the following rules at all:
          * fc_build_auxiliary_coordinate_latitude
          * fc_build_auxiliary_coordinate_latitude_rotated
          * fc_build_auxiliary_coordinate_longitude
          * fc_build_auxiliary_coordinate_longitude_rotated
    3. odd extras : Ukmo-attributes ; label-coords ; cell-measures; ancillary-variables
  * Later (intended) stages : 
     1. get all functionality tested + working with existing (Pyke) code
     2. add actions missing for TODO features
     3. get all tests pasing with actions implementation
     4. check that _all_ Iris tests run with temporary shim : do all netcdf loads _both_ ways + check results are the same
     5. remove the Pyke implementation (+dependency), and tidy away all the scaffolding

**NOTE**: I'm assuming that we will have enough confidence to "simply replace", which means
  1. everything goes via "new route"
  2. completely removed the old Pyke-based rules (and dependency)
  3. don't need, or have, any 'Future'-style switch for user to revert to 'old' mechanics

---
[Consult Iris pull request check list]( https://scitools-iris.readthedocs.io/en/latest/developers_guide/contributing_pull_request_checklist.html)
